### ISSUE TITLE
The code cannot run in version 1.55, but it can run in version 1.51

### ISSUE THREAD

**Issue by dunmengjun** [rust-lang/rust#87216]
Environment required:
> The code repo is https://github.com/dunmengjun/rCore-user.git

> Required qemu 5.0.0 (if setup from the source just need configure the riscv64 items, such as  ./configure --target-list=riscv64-softmmu,riscv64-linux-user)

>  Required rustup, after clone the code, run rustup show, it will setup the dependency automaticly.

When everything above is ok，you can run cargo build to build the elf file，and  use
```sh
qemu-riscv64 target/riscv64gc-unknown-none-elf/debug/02power
```
to run the elf program in qemu.

Currently the version of rust toolchain is the nightly newest (1.55), so you will get an error that
```
Segmentation fault
```

But if you change the toolchain to nightly-2021-01-30 (1.51), Just make this change, don’t change other things. the command will be running correctly

So the question is why it came out like this? Why I only changed the version of rust, but the results of running the same program are different?

And if we use the newest version of toolchain (1.55) and change the opt-level to 1 or 2 or 3 in .cargo/config, it will work.
```
[profile.dev]
opt-level = 1 # 1 or 2 or 3
```
but in nightly-2021-01-30 (1.51) it always work, don't need change anything.

Why is this so? Beg the big guys for help 😂

**Comment by Aaron1011** [comment]
This runs correctly for me under qemu-riscv64 6.0.0. Can you try updating your qemu version?

**Comment by dunmengjun** [comment]
@Aaron1011 thanks very mush，you provided me with ideas

In fact, qemu only needs to be upgraded to 5.2.0. because I have another lower-level program running on bare metal, this program is guided by rustsbi, and rustsbi currently does not support 6.0.0.

Thank you again for providing me with ideas

For the question, i gu, my guess is that qemu5.0.0 has a bug, which was fixed in 5.2.0. However, rust 1.55 must have changed some things to trigger the qemu5.0.0 bug.

Anyway, this must be a mess, haha
